src/neo/Network/P2P/TaskSession.cs

```diff
@@ -16,16 +16,23 @@ internal class TaskSession
 
         public bool HasTask => Tasks.Count > 0;
         public uint StartHeight { get; }
+        public bool IsFullNode { get; }
         public uint LastBlockIndex { get; set; }
 
         public TaskSession(IActorRef node, VersionPayload version)
         {
+            var fullNode = version.Capabilities.OfType<FullNodeCapability>().FirstOrDefault();
+
+            this.IsFullNode = fullNode != null;
             this.RemoteNode = node;
             this.Version = version;
-            this.StartHeight = version.Capabilities
-                .OfType<FullNodeCapability>()
-                .FirstOrDefault()?.StartHeight ?? 0;
+            this.StartHeight = fullNode?.StartHeight ?? 0;
             this.LastBlockIndex = this.StartHeight;
+
+            if (IsFullNode)
```

Comment:
```suggestion
            if (IsFullNode && Blockchain.Singleton.HeaderHeight >= this.LastBlockIndex)
```

Reply:
@shargon  Without this line, both side will add the `MemPoolTaskHash` in task manager, and this will block the belower node, as `session.AvailableTasks > 0`.

https://github.com/neo-project/neo/blob/a04722dfddeed406b915019ed8f93150fda4013d/src/neo/Network/P2P/TaskManager.cs#L227-L233

Reply:
Then a node never will receive the memory pool if it need to synchronize.

Reply:
This condition it's now before consume the task

https://github.com/neo-project/neo/pull/1544/files#diff-3aeec330e87769d041be8e902ffdc9b1R271

Reply:
> Then a node never will receive the memory pool if it need to synchronize.


?
The higher node will send the `MessageCommand.Mempool`.
The belower node will sync header and block first.

Reply:
I tested this code and the result is pass.
1. The node successfully share the mempool after the sync process is done.
2. Its block header goes up as soon as it connects to other nodes.
3. No exception thrown during the sync process.

Reply:
> The belower node will sync header and block first and also can send the mempool txs.

No, the belower node require a reconnection.

Reply:
I may not get your point, I think the below node doesn't need a reconnection, it's already connected. And as we can see, the higher node decide to whether to sync the mempool txs, the below node just receives p2p message is enough.

Reply:
If don't want to add this condition, we should remove `MemPoolTaskHash`  from the `hashes`.

https://github.com/neo-project/neo/blob/a04722dfddeed406b915019ed8f93150fda4013d/src/neo/Network/P2P/TaskManager.cs#L232-L233

Reply:
if we add this condition, we never add the flag of 'you need to sync your pool', then, as you said the higher node will sync the pool, but not the lower one, and this flag it's created only on new session, so the lower node won't send his tx never. Remember than both nodes need to sync his mempool, not only one

Reply:
But if both want to sync its mempool, both need to meet the condition. Which means, A's height >= B's height and B's height >=  A's height.

https://github.com/neo-project/neo/blob/3de164622b59f68f617a35f186daf3ac99741a1f/src/neo/Network/P2P/TaskManager.cs#L271-L273

Reply:
Two solutions:
1. Only allow the higher node to sync mempool, if so, we need to add this `Blockchain.Singleton.HeaderHeight >= this.LastBlockIndex` condition in `TaskSession`.
2. Allow both sides to sync mempool, if so, we need to filter the `MemPoolTaskHash` in `TaskManager`.  Add one line `hashes.Remove(MemPoolTaskHash)` after L232
https://github.com/neo-project/neo/blob/a04722dfddeed406b915019ed8f93150fda4013d/src/neo/Network/P2P/TaskManager.cs#L232-L233

Both for me are good. There maybe other way to fix this bug.

Reply:
I prefer the solution 2, could you push the path?

Reply:
Okay, I'll ask @nicolegys to help test it and then send the fixed pr.